### ISSUE TITLE
Could a setting be considered for SA1501/SA1503 to relax on some kind of statements?

### ISSUE THREAD

**Issue by liquidsnk** [DotNetAnalyzers/StyleCopAnalyzers#2252]
Somewhat related to #1175 our team uses something in the lines of the following rule to account for common guards:

> When the body of an if statement consists of a single jump statement (e.g. return, throw, break, continue) we allow that statement without braces and on the same line.

Which would mean that the initial code on that linked issued will be valid for us:

```csharp
if (condition) return;
```
**But not so** other kinds of statements like:

```csharp
if (condition) someVariable = "someValue";
```
or even
```csharp
if (condition) 
    someMethod();
```
We feel this is terse and removes clutter without causing readability concerns since the situation is pretty clear with jump statements which break the flow of execution (basically it allows us to have multiple one-liner guards, taking less space).

Now, **I know** that it's impossible to account for every team particular requirement or nuance and that maybe this is not even in the spirit of StyleCop at all (cannot say, while we feel we follow common sense standards, we have never adopted the tool until we discovered this great version using analyzers) but having the option to allow these kind of statement while also enforcing the already existing related restrictions is something that will be very good for us.

I imagined that before going and doing our own thing to enforce it or change the way we work to accomodate, it wouldn't hurt asking the project and the community if they would be interested in considering this behavior as optional or maybe even allow it as a contribution.

**Comment by vweijsters** [comment]
Thanks for the well written proposal. I see benefit in this (narrow) case so a 👍 from me. This will have to be optional, with the default behavior to remain the same as it is now.

The statement on the same line is paramount here, as it reduces the chance of somebody typing a next statement and assuming that it will be executed as part of the if statement. This could be further strengthened by demanding an empty line, but that should probably be a new rule.

**Comment by sharwell** [comment]
I believe the selective omission of braces where the child element is on the next line is covered by leaving SA1519 and SA1520 enabled, but disabling SA1503.

I'm :-1: on an additional option allowing the child statement to appear on the same line in these cases, as that option would be equivalent to the current ability to disable SA1501 in essentially all real-world cases.

**Comment by vweijsters** [comment]
@sharwell I do not agree that disabling SA1501 would cover all cases. This proposal talks about allowing this only for `if` statements. Disabling SA1501 would also open up other statements (like `lock`, `for`, `while`, etc.) to become single line statements, something that was never the intention of the proposal. 

An alternative for the proposed solution could be to only allow configuration of SA1501 so that it will allow single line `if` statements. The braces will then be mandatory, as SA1503 is not modified.

**Comment by sharwell** [comment]
@vweijsters Thanks, I misread a section of the proposal and was under the impression that the following was allowed:

```csharp
if (something)
  someMethod();
```

It seems like there are two options at play here:

1. May braces be omitted for an `if` statement with a single child that branches?
2. If so, is that single child placed on the current line or the following line?

**Comment by vweijsters** [comment]
I have no real preference for question 1.
I'd say that the answer to the 2nd question should be yes. 

💭 There may be benefit in adding it for people that cannot / may not upgrade to Visual Studio 2017 in the near future.

**Comment by liquidsnk** [comment]
Thanks for the thoughtful discussion guys. As mentioned, we only allow this for if statements with jumps as their single child, and we do like the current validations for **both** SA1501 and SA1503 enabled to happen on other cases than these. 

I also agree that those seem to be the options at play here and of course our answers would be "yes" and "current line", but were you to decide to go forward with this I'm not imagining any other way than having these as settings affecting the two rules, which I realize seems rather specific.

**Comment by GregReddick** [comment]
I'm going to give a mild 👎 . If implemented, it should be disabled by default. I think returns or throws should be the first thing on a line and be by themselves on a line to make them easier to find. I say this in spite of recognizing that it is a pretty frequent coding style to do validation and return at the start of a method, that there may be several of these, and that it extends quite a bit of vertical space when that is done.

I ask another question...what is the blank line spacing around the "if" if there are no braces. Is it legit to do:

if (condition1) return;
if (condition2) return;

or do there need to be blank lines between them. This may just open a whole new can of worms that shouldn't be opened.

**Comment by remyvd** [comment]
In principle I like this proposal. I always allow this in my guidelines, but with StyleCop it currently all or nothing in this case (SA1503). I want this fine grain ment of this rule, so a big YES ( 👍 ) to question 1.

For question 2 (same or following line) I see that most comments are answering this with 'current line'. I understand this answer, because it saves a line and looks clear on simple examples.

But after using both _current line_ and _following line_ in the past, I ended up with the _following line_ as preference, because of two reasons:

1. Whenever you want to place breakpoint for debugging, you can't place this only on the jump statement (return, throw, continue). When you are only interested in the 1 out of 1000 case that it throws, your will be glad you put a breakpoint only on the throw statement.

2. With conditions that are longer and in variable sizes it becomes more unclear what will happen when the condition is meet. Your eyes need to scan the line for the jump statement. If the jump statement is on the following line the logic is easier to spot. Also with 'current line' the line length more often exceeds my preferred max line length (130 characters).  See the following more real life example:

**Current line**
```C#
if (resource != "products" && resource != "people" && resource != "pricegroups") throw new NotImplementedException("Resource is not implemented yet!");
if (-99999 <= anotherlimit && anotherlimit <= 99999) throw new ArgumentOutOfRangeException("limit", "Must be between -99999 and 99999!");
if (limit > 100) throw new ArgumentOutOfRangeException("limit", "Must be less or equal to 100!");
if (obj == null) throw new ArgumentNullException("obj");
if (obj.Whatever == null) return;

foreach (var item in obj.Whatever.List)
{
    if (!item.Condition && someValue == otherValue && value1 == 2) continue;

    Process(item);
}
```

**Following line**
```C#
if (resource != "products" && resource != "people" && resource != "pricegroups")
    throw new NotImplementedException("Resource is not implemented yet!");
if (-99999 <= anotherlimit && anotherlimit <= 99999)
    throw new ArgumentOutOfRangeException("limit", "Must be between -99999 and 99999!");
if (limit > 100)
    throw new ArgumentOutOfRangeException("limit", "Must be less or equal to 100!");
if (obj == null)
    throw new ArgumentNullException("obj");
if (obj.Whatever == null)
    return;

foreach (var item in obj.Whatever.List)
{
    if (!item.Condition && someValue == otherValue && value1 == 2)
        continue;

    Process(item);
}
```

**Comment by sharwell** [comment]
:memo: I didn't have a question 1 and question 2. I was saying if implemented, there would be *two* new options in **stylecop.json** to allow each project to "answer the questions" with respect to just that project.

**Comment by remyvd** [comment]
@sharwell Aha, didn't catch that :-). Then yes, those would be the two new options to make SA1503 not too strict.

I'm personally in favor to add those options, because I always end op disable the stylecop rule and have a bad feeling about it, because SA1503 in itself is a good rule to have.

**Comment by buckleyGI** [comment]
My 2 cents as this is dear to me as well

```
// Allowed scenario 1: terse and **risk free** code
if (false) Console.WriteLine("statement1");

// Disallowed scenario 2: this is a trap for a future developer
if (false) 
	Console.WriteLine("statement2");

// Disallowed scenario 3: the trap is stepped on as the developer thinks statement4 is part of the if
if (false)
	Console.WriteLine("statement3");
	Console.WriteLine("statement4");
```

In the code snippets above, only the string "statement4" is printed out

I love to use scenario 1, **"the 1-liner"**, but can't because there is no way I will disable SA1503BracesMustNotBeOmitted

As soon as it is not a 1-liner anymore you one should introduce braces.
Otherwise a developer that sees scenario 2 can add a statement at the same indentation thinking it is part of the if (which it is not).
The following avoids the trap by adding braces to prevent the trap

```
// Allowed scenario 4
if (false){
	Console.WriteLine("statement3");
}
```

**Comment by sharwell** [comment]
@buckleyGI Your disallowed scenario 3 is impossible when [SA1137](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1137.md) is enabled. It should also be impossible when the built-in FormattingAnalyzer is enabled (IDE0055).

**Comment by buckleyGI** [comment]
@sharwell  Thanks, I didn't know about IDE0055 and it seems a solution if we can't enforce it.

But The version my company uses, VS 2022 Version 17.2.1, still allows me to setup the developer trap as shown in the screenshot below.

All the scenario's are allowed by this VS (no squiggles) and our Azure DevOps build server also doesn't complain, it just compiles.
Do we have to explicitly turn on IDE0055?

![image](https://user-images.githubusercontent.com/2102864/183602380-7f732f02-87be-46d4-b1f3-e73ca4ca8315.png)

It is not 100% clear how these IDExxx rules work. Are they rules only for the IDE (suggestion/quick fixes) or can you also enforce them at compilation time?

Is it because we still use ruleset instead of the editorconfig and should we migrate to editorconfig to enforce IDE055 by DevOps? I read in the documentation that you can influence the IDExxx rules only with editorconfig.

**Comment by buckleyGI** [comment]
@sharwell Thanks, I didn't know about SA1137 and the following configuration will allow the one liner without introducing the trap.

```
    <Rule Id="SA1503" Action="None" />
    <Rule Id="SA1137" Action="Error" />
```
